sweep/tx_input_set.go

```diff
@@ -540,3 +540,194 @@ func createWalletTxInput(utxo *lnwallet.Utxo) (input.Input, error) {
 		&utxo.OutPoint, witnessType, signDesc, heightHint,
 	), nil
 }
+
+// budgetInputSet implements the interface `InputSet`. It takes a list of
+// pending inputs which share the same deadline height and groups them into a
+// set conditionally based on their economical values.
+type budgetInputSet struct {
+	// inputs is the set of inputs that have been added to the set after
+	// considering their economical contribution.
+	inputs []*pendingInput
+
+	// budget is the total budget that can be spent as fees.
+	budget btcutil.Amount
+
+	// deadlineHeight is the height which the inputs in this set must be
+	// confirmed by.
+	deadlineHeight int32
+}
+
+// Compile-time constraint to ensure budgetInputSet implements InputSet.
+var _ InputSet = (*budgetInputSet)(nil)
+
+// newBudgetInputSet creates a new budgetInputSet.
+func newBudgetInputSet(deadlineHeight int32) *budgetInputSet {
+	return &budgetInputSet{
+		deadlineHeight: deadlineHeight,
+	}
+}
+
+// addInput adds an input to the set if it is economically viable. It returns
+// true if the input was added to the set.
+func (b *budgetInputSet) addInput(pi *pendingInput) bool {
+	// If the input comes with a required tx out that is below dust, we
+	// won't add it.
+	//
+	// NOTE: only HtlcSecondLevelAnchorInput returns non-nil RequiredTxOut.
+	reqOut := pi.RequiredTxOut()
+	if reqOut != nil {
+		// Fetch the dust limit for this output.
+		dustLimit := lnwallet.DustLimitForSize(len(reqOut.PkScript))
+		if btcutil.Amount(reqOut.Value) < dustLimit {
+			log.Errorf("Rejected input=%v due to dust required "+
+				"output=%v, limit=%v", pi, reqOut.Value,
+				dustLimit)
+
+			return false
+		}
+	}
+
+	// Make sure the deadlines can be matched to make sure the fee bumper
+	// can given accurate feerates. More importantly, for second-level HTLC
+	// txns, the same tx-level `nLockTime` must be the same in order for
+	// them to be aggregated into the same sweeping tx. This `nLockTime`
+	// value is implicitly expressed as the deadline height.
+	if b.deadlineHeight != pi.params.DeadlineHeight {
```

Comment:
I suppose that includes the ss-htlc txs too... 😕

Reply:
So I came up with a way to solve the issue, with the help of `fn.Option`. It goes like this,
- for ss-htlcs, they must specify deadlines to `fn.Some`, with their `nLocktime` values.
- for non-time-sensitive outputs, they specify their deadlines to `fn.None`.
- when aggregating, all `fn.Nonce` will be put in one set, allowing us to sweep them in one tx.

This can be extended to allow adding a later deadline input, as long as it's `fn.None`.